internal/util/debug/debug.go

```diff
@@ -97,25 +123,98 @@ func Listen(opts *ListenOpts) (*Handler, error) {
 		}),
 	))
 
+	http.HandleFunc("/debug/archive", promhttp.InstrumentHandlerDuration(
+		archiveDurations,
+		http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			rw.Header().Set("Content-Type", "application/zip")
+
+			rw.Header().Set(
+				"Content-Disposition",
+				fmt.Sprintf("attachment; filename=FerretDB-debug-%d.zip", time.Now().UnixMilli()),
+			)
+
+			ctx := req.Context()
+			zipWriter := zip.NewWriter(rw)
+
+			defer func() {
+				if err := zipWriter.Close(); err != nil {
+					opts.L.Error("Archive handler failed", zap.Error(err))
+					rw.WriteHeader(http.StatusInternalServerError)
+
+					return
+				}
+			}()
+
+			metricsFile, err := zipWriter.Create("metrics")
+			if err != nil {
+				opts.L.Error("Archive handler failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			// we use *http.Request instead of http.Get function to provide the ctx
+			scrapeReq := must.NotFail(http.NewRequestWithContext(
+				ctx,
+				http.MethodGet,
+				fmt.Sprintf("http://%s%s", req.Host, "/debug/metrics"),
```

Comment:
Why not use https://pkg.go.dev/net/url#JoinPath?

Reply:
We don't handle user provided paths here, path traversal is not a problem here.
We can use `JoinPath` here nevertheless, but the code'll be just less simple:
```
			must.NotFail(url.JoinPath("http://"+req.Host, "/debug/metrics")),
```

Although I think what could be done is to not use `req.Host` at all. It doesn't make sense to use and potentially might be harmful, thank you for noticing the line!

Reply:
Ok. What is harmful though?

Reply:
The `Host` field is not validated by the server, and it can be manipulated by the client. That might lead to things such as forcing the server to query user-specified server, or something unexpected and more harmful.

Reply:
🏅